Use GCC for LD on MIPS processors

rm5248 opened the pull request:
## Summary
I built GCC 11.2 with crosstool-ng and found that the linking process failed.  This led me to issue #3826, which is the same issue but for ARM.  I applied effectively the same patch as #3836 in order to get it to compile

## Impact
Seems to be minimal.  Other compilers may have issues.

## Testing
Nuttx now compiles with newer versions of GCC.  This shouldn't have any effect on the actual compiled code.  I have not tested with other compilers apart from the one crosstool-ng built one.

xiaoxiang781216 commented:
@rm5248 there are some compile errors. Please reference the last arm Toolchain.defs too.

acassis commented:
@rm5248 very nice!!! What board did you use for test? How did you generate the toolchain? I tried this way https://acassis.wordpress.com/2022/08/06/using-crosstool-ng-to-create-mips-toolchain/ but it didn't create a mips-unknown-elf-gcc! My goal is to get NuttX running on Dmitry visit card: https://dmitry.gr/?r=05.Projects&proj=33.%20LinuxCard and on this uMIPS emulador.

rm5248 commented:
@xiaoxiang781216 I did notice an error and fixed it, it should be good now.  The build still seems to be failing, but that appears to be a different issue with the compiler.

rm5248 commented:
@acassis Based off of [this mailing list post](https://lists.apache.org/thread/193jmh192zm157sk6pmsglrkf6b3kvpw), I just started with the mips-unknown-elf toolchain and modified it slightly to build for the PIC32

xiaoxiang781216 commented:
> @xiaoxiang781216 I did notice an error and fixed it, it should be good now. The build still seems to be failing, but that appears to be a different issue with the compiler.

Look like the endian of compiled generated code is different from the expectation of link script. Please try the different value of OUTPUT_FORMAT in the link script.

acassis commented:
> > @xiaoxiang781216 I did notice an error and fixed it, it should be good now. The build still seems to be failing, but that appears to be a different issue with the compiler.
> 
> Look like that the endian of compiled generated code is different from the expectation of link script. Please try the different value of OUTPUT_FORMAT in the link script.

Yes, the mips-unknown-elf toolchain is big-endian by default, I needed to change it to little-endian

acassis commented:
> @acassis Based off of [this mailing list post](https://lists.apache.org/thread/193jmh192zm157sk6pmsglrkf6b3kvpw), I just started with the mips-unknown-elf toolchain and modified it slightly to build for the PIC32

What did you do after "ct-ng build" ?

rm5248 commented:
> > @acassis Based off of [this mailing list post](https://lists.apache.org/thread/193jmh192zm157sk6pmsglrkf6b3kvpw), I just started with the mips-unknown-elf toolchain and modified it slightly to build for the PIC32
> 
> What did you do after "ct-ng build" ?

Added it to the PATH and set CROSSDEV:

```
export CROSSDEV=mipsel-elf-
export PATH=/home/robert/x-tools/mipsel-elf/bin:$PATH
```

Here's the output from `ct-ng savedefconfig`:
```
CT_CONFIG_VERSION="4"
CT_ARCH_MIPS=y
CT_OMIT_TARGET_VENDOR=y
CT_ARCH_LE=y
CT_ARCH_FLOAT_SW=y
CT_STATIC_TOOLCHAIN=y
CT_LIBC_NONE=y
CT_CC_GCC_mips_plt=y
CT_CC_LANG_CXX=y
CT_DEBUG_GDB=y
CT_ZLIB_NEEDED=y
```

acassis commented:
Thank you @rm5248 !! @xiaoxiang781216 maybe the issue is caused by incompatibility with pinguino after this change. But since Robert is also using little endian, but issue could be specific of pinguino

rm5248 commented:
As far as I am aware, all of the PIC32 processors are always little-endian, so it seems that if there is a problem with the big vs. little endian that is a problem with the compiler that is used and/or the configuration of nuttx.

btashton commented:
At least a while back I ran some testing under qemu. But looks like you might have also been on that thread. 

https://lists.apache.org/thread/hk8gdtr495qbl2fg4yx1gfotm319npmy

pkarashchenko commented on arch/mips/src/mips32/Toolchain.defs:
```diff
@@ -245,14 +245,14 @@ endif
 
 ifeq ($(CONFIG_DEBUG_OPT_UNUSED_SECTIONS),y)
-  LDFLAGS          += --gc-sections
+  LDFLAGS          += -Wl,--gc-sections
```
I think there was an automated way of adding -Wl in case if gcc is used for linking in case of ARM. Maybe the same approach can be used

xiaoxiang781216 replied:
Yes, @rm5248 you can reference how arm support link with gcc and ld at the same time.

rm5248 replied:
It looks like the commit that I was basing this off of was reverted at some point, so I will configure this differently.

xiaoxiang781216 replied:
Yes, you need check the last code.

rm5248 commented:
after looking at this more, everything builds fine for me with the current master branch.  I think I was in some weird state where I had either tried to use two different compilers and/or I had leftover data from something old.

Closing as this is not an issue.
